New .NET MAUI Blazor Hybrid template (#32793)
dotnet/AspNetCore.Docs: aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md

For prerequisites and preliminary steps, see <xref:blazor/hybrid/tutorials/maui>. We recommend using the .NET MAUI Blazor Hybrid tutorial to set up your local system for .NET MAUI development before using the guidance in this article.

:::moniker range=">= aspnetcore-9.0"

> ## .NET MAUI Blazor Web App project template

Reviewer: ```suggestion
## .NET MAUI Blazor Hybrid and Web App solution template
```

> The .NET MAUI Blazor Web App project template allows you to choose a Blazor interactive render mode for the web app and creates the appropriate projects for the app, including a Blazor Web App and a .NET MAUI Blazor Hybrid app. A shared Razor class library (RCL) maintains the Razor components for the app's UI. The template also provides sample code to shows you how to use dependency injection to provide different interface implementations for the Blazor Hybrid and Blazor Web App, which is covered in the [Using interfaces to support different device implementations](#using-interfaces-to-support-different-device-implementations) section of this article.

Reviewer: ```suggestion
The .NET MAUI Blazor Hybrid and Web App solution template sets up a solution that targets Android, iOS, Mac, Windows and Web that reuses UI. You can choose a Blazor interactive render mode for the web app and it creates the appropriate projects for the app, including a Blazor Web App and a .NET MAUI Blazor Hybrid app. A shared Razor class library (RCL) maintains the Razor components for the app's UI. The template also provides sample code to show you how to use dependency injection to provide different interface implementations for the Blazor Hybrid and Blazor Web App, which is covered in the [Using interfaces to support different device implementations](#using-interfaces-to-support-different-device-implementations) section of this article.
```

Create an app from the project template with the following .NET CLI command:

```dotnetcli
dotnet new maui-blazor-web -o MauiBlazorWeb -int Server  -ai
```

In the preceding command:

* The `-o|--output` option creates a new folder for the app named `MauiBlazorWeb`.
* The `-int|--interactivity` option sets up the interactivity location to `Server`.
* The `-ai|--all-interactive` option specifies global interactivity, which is important because MAUI apps always run interactively and throw errors on Razor component pages that explicitly specify a render mode. If you don't use a global render mode, you must implement the approach described in the [Use Blazor render modes](#use-blazor-render-modes) section. For more information, see [BlazorWebView needs a way to enable overriding ResolveComponentForRenderMode (`dotnet/aspnetcore` #51235)](https://github.com/dotnet/aspnetcore/issues/51235).

<!-- UPDATE 9.0 Provide the project template's name here for VS, possibly
                using a tooling pivot for the article. -->

:::moniker-end

:::moniker range="< aspnetcore-9.0"

> ## .NET MAUI Blazor Web App sample app

Reviewer: ```suggestion
## .NET MAUI Blazor Hybrid and Web App sample app
```

Reviewer: The only difference between the guidance below and what the solution template creates for you is that the name of the MAUI project does not have ".Maui" naming, it's just the name that the user indicated as the name of their project/solution. So if you named it "MyApp" it would be: 
  * Solution: `MyApp`
  * MAUI project: `MyApp`
  * Blazor Web App: `MyApp.Web`
  * Blazor Web Assembly App: `MyApp.Web.Client`
  * Razor class library (RCL): `MyApp.Shared`

The following example demonstrates how to use an interface to call into different implementations across the web app and the native (MAUI) app. The following example creates a component that displays the device form factor. Use the MAUI abstraction layer for native apps and provide an implementation for the web app.

In the Razor class library (RCL), an `Interfaces` folder contains an `IFormFactor` interface.

`Interfaces/IFormFactor.cs`:

:::code language="csharp" source="~/../blazor-samples/8.0/MauiBlazorWeb/MauiBlazorWeb.Shared/Interfaces/IFormFactor.cs":::

The following `DeviceFormFactor` component is present in the RCL's `Components` folder:

`Components/Pages/DeviceFormFactor.razor`:

:::code language="razor" source="~/../blazor-samples/8.0/MauiBlazorWeb/MauiBlazorWeb.Shared/Components/Pages/DeviceFormFactor.razor":::

> In the RCL, an entry for the `DeviceFormFactor` component is part of the navigation menu in the `NavMenu` component.

Reviewer: In the new solution template, the form factor UI is on the home page. There isn't a DeviceFormFactor.razor page and there is no change to the NavMenu.

</div>
```

The web and native apps contain the implementations for `IFormFactor`.

In the Blazor Web App, a folder named `Services` contains the following `FormFactor.cs` file with the `FormFactor` implementation for web app use.

`Services/FormFactor.cs` (Blazor Web App project):

:::code language="csharp" source="~/../blazor-samples/8.0/MauiBlazorWeb/MauiBlazorWeb.Web/Services/FormFactor.cs":::

In the MAUI project, a folder named `Services` contains the following `FormFactor.cs` file with the `FormFactor` implementation for native use. The MAUI abstractions layer is used to write code that works on all native device platforms.

`Services/FormFactor.cs` (MAUI project):
 
:::code language="csharp" source="~/../blazor-samples/8.0/MauiBlazorWeb/MauiBlazorWeb.Maui/Services/FormFactor.cs":::

Dependency injection is used to obtain the implementations of these services.

In the MAUI project, the `MauiProgram.cs` file has following `using` statements at the top of the file:

```csharp

> using MauiBlazorWeb.Maui.Services;

Reviewer: ```suggestion
using MauiBlazorWeb.Services;
```
